crates/extensions/tedge_mqtt_bridge/src/lib.rs

```diff
@@ -0,0 +1,202 @@
+use tedge_config::TEdgeConfig;
+
+pub struct MqttBridgeActorBuilder {
+    signal_sender: mpsc::Sender<RuntimeRequest>,
+}
+
+impl MqttBridgeActorBuilder {
+    pub async fn new(tedge_config: &TEdgeConfig, cloud_topics: &[impl AsRef<str>]) -> Self {
+        let tls_config = create_tls_config(
+            tedge_config.c8y.root_cert_path.clone().into(),
+            tedge_config.device.key_path.clone().into(),
+            tedge_config.device.cert_path.clone().into(),
+        )
+        .unwrap();
+        let (signal_sender, _signal_receiver) = mpsc::channel(10);
+
+        let prefix = tedge_config.c8y.bridge.topic_prefix.clone();
+        let mut local_config = MqttOptions::new(
+            format!("tedge-mapper-bridge-{prefix}"),
+            &tedge_config.mqtt.client.host,
+            tedge_config.mqtt.client.port.into(),
+        );
+        // TODO cope with secured mosquitto
+        local_config.set_manual_acks(true);
+        let mut cloud_config = MqttOptions::new(
+            tedge_config.device.id.try_read(tedge_config).unwrap(),
+            tedge_config
+                .c8y
+                .mqtt
+                .or_config_not_set()
+                .unwrap()
+                .host()
+                .to_string(),
+            8883,
+        );
+        cloud_config.set_manual_acks(true);
+
+        cloud_config.set_transport(Transport::tls_with_config(tls_config.into()));
+        let topic_prefix = format!("{prefix}/");
+        let (local_client, local_event_loop) = AsyncClient::new(local_config, 10);
+        let (cloud_client, cloud_event_loop) = AsyncClient::new(cloud_config, 10);
+
+        local_client
+            .subscribe(format!("{topic_prefix}#"), QoS::AtLeastOnce)
+            .await
+            .unwrap();
+        for topic in cloud_topics {
+            cloud_client
+                .subscribe(topic.as_ref(), QoS::AtLeastOnce)
+                .await
+                .unwrap();
+        }
+
+        let (tx_pubs_from_cloud, rx_pubs_from_cloud) = mpsc::channel(10);
+        let (tx_pubs_from_local, rx_pubs_from_local) = mpsc::channel(10);
+        tokio::spawn(one_way_bridge(
+            local_event_loop,
+            cloud_client,
+            move |topic| topic.strip_prefix(&topic_prefix).unwrap().into(),
+            tx_pubs_from_local,
+            rx_pubs_from_cloud,
+        ));
+        tokio::spawn(one_way_bridge(
+            cloud_event_loop,
+            local_client,
+            move |topic| format!("{prefix}/{topic}").into(),
+            tx_pubs_from_cloud,
+            rx_pubs_from_local,
+        ));
+
+        Self { signal_sender }
+    }
+
+    pub(crate) fn build_actor(self) -> MqttBridgeActor {
+        MqttBridgeActor {}
+    }
+}
+
+async fn one_way_bridge<F: for<'a> Fn(&'a str) -> Cow<'a, str>>(
+    mut recv_event_loop: EventLoop,
+    target: AsyncClient,
+    transform_topic: F,
+    mut tx_pubs: mpsc::Sender<Publish>,
+    mut rx_pubs: mpsc::Receiver<Publish>,
+) {
```

Comment:
There was some equivalent to the assertion with a call to `unwrap()` when  calling `forward_pkid_to_received_msg.remove()`. By unwrapping the value, we ensure the packet id is known, and this should only happen reliably if everything is working correctly. If we have accidentally read packet ids for the cloud when we should be reading them for the local connection, one of the directions will quickly encounter an acknowledgement for an unknown packet id. I've currently removed this since I'm confident the loop now works, and I think this particular bug is difficult to achieve due to the required input to the function.

I've also modified the channels to be a single bidirectional channel to avoid exposing states which don't make sense, and (mostly) documented the function (still a little bit to do there, and ascii art will likely help)

Reply:
> I've also modified the channels to be a single bidirectional channel to avoid exposing states which don't make sense

This is indeed a good idea.

Reply:
Here is some ascii art describing how the bridge works https://gist.github.com/didier-wenzek/d37d7cdb8bdbd5b0b572f6fb81d42801